types/jasmine/index.d.ts

```diff
@@ -15,7 +15,7 @@
 // TypeScript Version: 2.8
 // For ddescribe / iit use : https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/karma-jasmine/karma-jasmine.d.ts
 
-type ImplementationCallback = (() => Promise<any>) | ((done: DoneFn) => void);
+type ImplementationCallback = (() => PromiseLike<any>) | ((done: DoneFn) => void) | (() => void);
```

Comment:
Nit: you can do (done?: ...)

Reply:
Had some problems there, but will try.

Reply:
Ah, I remember the problem is that tsc@3.3.3333 complains

```
jasmine-tests.ts:1026:19 - error TS2345: Argument of type '(done: DoneFn) => void' is not assignable to parameter of type 'ImplementationCallback'.
  Types of parameters 'done' and 'done' are incompatible.
    Type 'DoneFn | undefined' is not assignable to type 'DoneFn'.
      Type 'undefined' is not assignable to type 'DoneFn'.

1026         afterEach((done: DoneFn) => {
```

for an ImplementationCallback declaration like so

```
type ImplementationCallback = (() => PromiseLike<any>) | ((done?: DoneFn) => void);
```

Given the original declaration 

```
type ImplementationCallback = (() => PromiseLike<any>) | ((done: DoneFn) => void) | (() => void);
```

the tests are compiled just fine except for

```
jasmine-tests.ts:122:29 - error TS2345: Argument of type 'ObjectContaining<{ a: string; }>' is not assignable to parameter of type 'Expected<{ a: string; b?: undefined; } | { b: string; a?: undefined; }>'.
  Type 'ObjectContaining<{ a: string; }>' is not assignable to type 'ObjectContaining<{ a: string; b?: undefined; } | { b: string; a?: undefined; }>'.
    Types of parameters 'sample' and 'sample' are incompatible.
      Type 'Partial<{ a: string; b?: undefined; }> | Partial<{ b: string; a?: undefined; }>' is not assignable to type 'Partial<{ a: string; }>'.
        Type 'Partial<{ b: string; a?: undefined; }>' is not assignable to type 'Partial<{ a: string; }>'.
          Type '{ b: string; a?: undefined; }' is not assignable to type '{ a: string; }'.

122         expect(a).toContain(jasmine.objectContaining({ a: "foo" }));
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

jasmine-tests.ts:861:33 - error TS2345: Argument of type 'ObjectContaining<{ a: number; foo: number; }>' is not assignable to parameter of type 'Expected<fooType>'.
  Type 'ObjectContaining<{ a: number; foo: number; }>' is not assignable to type 'ObjectContaining<fooType>'.
    Types of parameters 'sample' and 'sample' are incompatible.
      Type 'Partial<fooType>' is not assignable to type 'Partial<{ a: number; foo: number; }>'.
        Property 'foo' is missing in type 'fooType' but required in type '{ a: number; foo: number; }'.

861         expect(foo).not.toEqual(jasmine.objectContaining({
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~
862             a: 37,
    ~~~~~~~~~~~~~~~~~~
...
864             // b: '123', <-- this would cause an error as `b` defined as number in fooType
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
865         }));
    ~~~~~~~~~~

  jasmine-tests.ts:863:13
    863             foo: 2, // <-- this does not cause an error as the compiler cannot infer the type completely
                    ~~~~~~
    'foo' is declared here.
```

with strictFunctionTypes: true.

Reply:
Oh that makes sense since the overload here is treated different than an optional parameter.